probably/pgcl/parser.py

```diff
@@ -313,30 +388,94 @@ def _parse_instrs(t: Tree) -> List[Instr]:
     return [_parse_instr(_as_tree(t)) for t in t.children]
 
 
-def _parse_program(t: Tree) -> Program:
+def _parse_queries(t: Tree) -> List[Instr]:
+    assert t.data == "queries"
+    return [_parse_query(_as_tree(t)) for t in t.children]
+
+
+def _parse_query(t: Tree):
+    if t.data == 'expectation':
+        return ExpectationInstr(_parse_expr(_child_tree(t, 0)))
+    elif t.data == 'prquery':
+        return ProbabilityQueryInstr(_parse_expr(_child_tree(t, 0)))
+    elif t.data == 'print':
+        return PrintInstr()
+    elif t.data == 'optimize':
+        mode = _parse_var(_child_tree(t, 2))
+        if mode == "MAX":
+            opt_type = OptimizationType.MAXIMIZE
+        elif mode == "MIN":
+            opt_type = OptimizationType.MINIMIZE
+        else:
+            raise SyntaxError(f"The Optimization can either be 'MAX' or 'MIN', but not {mode}")
+        parameter = _parse_var(_child_tree(t, 1))
+        if parameter not in parameters:
+            raise SyntaxError(
+                f"In Optimization queries, the variable can only be a program parameter, was {parameter}.")
+        return OptimizationQuery(_parse_expr(_child_tree(t, 0)), parameter, opt_type)
+    elif t.data == "plot":
+        if len(t.children) == 3:
+            lit = _parse_literal(_child_tree(t, 2))
+            if isinstance(lit, BoolLitExpr):
+                raise SyntaxError("Plot instructions cannot handle boolean literals as arguments")
```

Comment:
This should also be done in the type checker, not in the parser.

Reply:
In general, do you want to have language extensions which allow for usability? e.g. Printing intermediate results via !Print, or getting expected Values with ?Ex[expr], or Pooling some stuff with !Plot...?